File: experimental/ip-geofeed/internal/parser/parser.go
package parser

import (
	"bufio"
	"context"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// Row represents a single CSV row
type Row struct {
	Line        int
	IPPrefix    string
	CountryCode string
	RegionCode  string
	City        string
	PostalCode  string
}

// Record represents a single geofeed entry
type Record struct {
	Geofeed string `json:"geofeed"`
	Inetnum string `json:"inetnum"`
	Source  string `json:"source,omitempty"`
	Netname string `json:"netname,omitempty"`
	Country string `json:"country,omitempty"`
	Org     string `json:"org,omitempty"`
	AdminC  string `json:"admin-c,omitempty"`
	TechC   string `json:"tech-c,omitempty"`
	MntBy   string `json:"mnt-by,omitempty"`
	City    string `json:"city,omitempty"`
}

// Records is a collection of Record entries
type Records []Record

// EncodeInetnum encodes the inetnum field into a filesystem-friendly string
func (p *Record) EncodeInetnum() string {
	s := strings.TrimSpace(p.Inetnum)

	// Convert range separator first
	s = strings.ReplaceAll(s, " - ", "_to_")

	// Replace characters with hyphen
	s = strings.ReplaceAll(s, ".", "-")
	s = strings.ReplaceAll(s, ":", "-")
	s = strings.ReplaceAll(s, "/", "-")

	// Remove unwanted characters (but keep - and _)
	reg := regexp.MustCompile(`[^a-zA-Z0-9\-_]+`)
	s = reg.ReplaceAllString(s, "")

	return s
}

// LoadPublishers reads and unmarshals a publishers.json file
func LoadPublishers(filePath string) (Records, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read file: %w", err)
	}

	var publishers Records
	if err := json.Unmarshal(data, &publishers); err != nil {
		return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
	}

	return publishers, nil
}

// isURL checks if the given string is a valid HTTP/HTTPS URL
func isURL(fileSource string) bool {
	u, err := url.Parse(fileSource)
	return err == nil && (u.Scheme == "http" || u.Scheme == "https")
}

// downloadFile downloads a file from a URL and saves it to the run/data folder
func downloadFile(urlStr string) (string, error) {
	// Create run/data directory if it doesn't exist
	runDataDir := filepath.Join("run", "data")
	if err := os.MkdirAll(runDataDir, 0755); err != nil {
		return "", fmt.Errorf("creating run/data directory: %w", err)
	}

	// Extract filename from URL
	u, err := url.Parse(urlStr)
	if err != nil {
		return "", fmt.Errorf("parsing URL: %w", err)
	}
	filename := filepath.Base(u.Path)
	if filename == "" || filename == "/" || filename == "." {
		filename = "geofeed.csv"
	}

	// Full path where file will be saved
	outputPath := filepath.Join(runDataDir, filename)

	// Create request with headers
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, "GET", urlStr, nil)
	if err != nil {
		return "", fmt.Errorf("creating request: %w", err)
	}

	// req.Header.Set("User-Agent", "curl/8.0.0")
	req.Header.Set("User-Agent", "Mozilla/5.0 (compatible; IPGeofeed/1.0)")
	req.Header.Set("Accept", "*/*")

	client := &http.Client{}

	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("downloading file: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("download failed with status %d: %s", resp.StatusCode, urlStr)
	}

	// Create file
	out, err := os.Create(outputPath)
	if err != nil {
		return "", fmt.Errorf("creating file %s: %w", outputPath, err)
	}
	defer out.Close()

	// Copy content
	if _, err := io.Copy(out, resp.Body); err != nil {
		return "", fmt.Errorf("writing file: %w", err)
	}

	return outputPath, nil
}

// resolveFilePath resolves the actual file path, downloading from URL if necessary
func resolveFilePath(fileSource string) (string, error) {
	if isURL(fileSource) {
		fmt.Printf("Downloading from URL: %s\n", fileSource)
		return downloadFile(fileSource)
	}
	// Return local file path as-is
	return fileSource, nil
}

// ParseCSV reads and parses a CSV geofeed file from local path or URL
func ParseCSV(fileSource string, limit int) ([]Row, map[int]string, int, int, error) {
	// Resolve file path (download from URL if necessary)
	filepath, err := resolveFilePath(fileSource)
	if err != nil {
		return nil, nil, 0, 0, fmt.Errorf("resolving file path: %w", err)
	}

	file, err := os.Open(filepath)
	if err != nil {
		return nil, nil, 0, 0, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var rows []Row
	comments := make(map[int]string)
	lineNum := 0
	invalidEntries := 0
	validEntries := 0
	originalCSVCols := 0

	for scanner.Scan() {
		line := scanner.Text()
		trimmed := strings.TrimSpace(line)
		lineNum++

		// Store empty lines and comments
		if trimmed == "" {
			comments[lineNum] = line
			continue
		}
		if strings.HasPrefix(trimmed, "#") {
			comments[lineNum] = line
			continue
		}

		// Check if we've reached the limit
		if limit > 0 && validEntries >= limit {
			break
		}

		// Parse CSV row from raw line
		reader := csv.NewReader(strings.NewReader(line))
		reader.TrimLeadingSpace = true
		record, err := reader.Read()
		if err != nil {
			continue
		}

		// Accept rows with 1 to 5 columns, pad missing columns with empty strings
		if len(record) < 1 || len(record) > 5 {
			invalidEntries++
			continue
		}

		// Track the maximum number of columns seen across all valid rows
		if len(record) > originalCSVCols {
			originalCSVCols = len(record)
		}

		for len(record) < 4 {
			record = append(record, "")
		}

		row := Row{
			Line:        lineNum,
			IPPrefix:    strings.TrimSpace(record[0]),
			CountryCode: strings.TrimSpace(record[1]),
			RegionCode:  strings.TrimSpace(record[2]),
			City:        strings.TrimSpace(record[3]),
		}
		if len(record) == 5 {
			row.PostalCode = strings.TrimSpace(record[4])
		}
		rows = append(rows, row)
		validEntries++
	}

	if err := scanner.Err(); err != nil {
		return nil, nil, 0, 0, err
	}
	return rows, comments, invalidEntries, originalCSVCols, nil
}

File: experimental/ip-geofeed2/internal/parser/parser.py
import csv
import io
import json
import os
import re
import urllib.parse
import urllib.request
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class Row:
    """Represents a single CSV row."""
    line: int = 0
    ip_prefix: str = ""
    country_code: str = ""
    region_code: str = ""
    city: str = ""
    postal_code: str = ""


@dataclass
class Record:
    """Represents a single geofeed entry from publishers JSON."""
    geofeed: str = ""
    inetnum: str = ""
    source: str = ""
    netname: str = ""
    country: str = ""
    org: str = ""
    admin_c: str = ""
    tech_c: str = ""
    mnt_by: str = ""
    city: str = ""

    def encode_inetnum(self) -> str:
        """Encodes the inetnum field into a filesystem-friendly string."""
        s = self.inetnum.strip()
        s = s.replace(" - ", "_to_")
        s = s.replace(".", "-")
        s = s.replace(":", "-")
        s = s.replace("/", "-")
        s = re.sub(r"[^a-zA-Z0-9\-_]+", "", s)
        return s


def load_publishers(file_path: str) -> list[Record]:
    """Reads and unmarshals a publishers.json file."""
    with open(file_path, "r") as f:
        data = json.load(f)

    publishers = []
    for item in data:
        publishers.append(Record(
            geofeed=item.get("geofeed", ""),
            inetnum=item.get("inetnum", ""),
            source=item.get("source", ""),
            netname=item.get("netname", ""),
            country=item.get("country", ""),
            org=item.get("org", ""),
            admin_c=item.get("admin-c", ""),
            tech_c=item.get("tech-c", ""),
            mnt_by=item.get("mnt-by", ""),
            city=item.get("city", ""),
        ))
    return publishers


def _is_url(file_source: str) -> bool:
    """Checks if the given string is a valid HTTP/HTTPS URL."""
    parsed = urllib.parse.urlparse(file_source)
    return parsed.scheme in ("http", "https")


def _download_file(url_str: str) -> str:
    """Downloads a file from a URL and saves it to the run/data folder."""
    run_data_dir = os.path.join("run", "data")
    os.makedirs(run_data_dir, exist_ok=True)

    parsed = urllib.parse.urlparse(url_str)
    filename = os.path.basename(parsed.path)
    if not filename or filename in ("/", "."):
        filename = "geofeed.csv"

    output_path = os.path.join(run_data_dir, filename)

    req = urllib.request.Request(url_str)
    req.add_header("User-Agent", "Mozilla/5.0 (compatible; IPGeofeed/1.0)")
    req.add_header("Accept", "*/*")

    with urllib.request.urlopen(req, timeout=120) as resp:
        if resp.status != 200:
            raise RuntimeError(f"Download failed with status {resp.status}: {url_str}")
        with open(output_path, "wb") as out:
            while True:
                chunk = resp.read(8192)
                if not chunk:
                    break
                out.write(chunk)

    return output_path


def _resolve_file_path(file_source: str) -> str:
    """Resolves the actual file path, downloading from URL if necessary."""
    if _is_url(file_source):
        print(f"Downloading from URL: {file_source}")
        return _download_file(file_source)
    return file_source


def parse_csv(file_source: str, limit: int = 0) -> tuple[list[Row], dict[int, str], int]:
    """Reads and parses a CSV geofeed file from local path or URL.

    Returns:
        Tuple of (rows, comments, invalid_entries)
    """
    filepath = _resolve_file_path(file_source)

    rows: list[Row] = []
    comments: dict[int, str] = {}
    line_num = 0
    invalid_entries = 0
    valid_entries = 0

    with open(filepath, "r", newline="") as f:
        for line in f:
            line = line.rstrip("\n").rstrip("\r")
            trimmed = line.strip()
            line_num += 1

            if trimmed == "":
                comments[line_num] = line
                continue
            if trimmed.startswith("#"):
                comments[line_num] = line
                continue

            if limit > 0 and valid_entries >= limit:
                break

            try:
                reader = csv.reader(io.StringIO(line), skipinitialspace=True)
                record = next(reader)
            except (csv.Error, StopIteration):
                continue

            if len(record) < 4 or len(record) > 5:
                invalid_entries += 1
                continue

            row = Row(
                line=line_num,
                ip_prefix=record[0].strip(),
                country_code=record[1].strip(),
                region_code=record[2].strip(),
                city=record[3].strip(),
            )
            if len(record) == 5:
                row.postal_code = record[4].strip()

            rows.append(row)
            valid_entries += 1

    return rows, comments, invalid_entries
